Step1 문자열 덧셈 계산기

Dev-RubinJo opened the pull request:
TDD 흐름에 따라 문자열 덧셈 계산기를 구현해봤습니다. 
첫번째로 요구 사항에 따라 구현해야할 객체들을 정리했습니다.
그 후에 컴파일이 안되더라도 일단 정리한 객체들에 대한 테스트 코드를 작성했습니다.
그리고 통과만 하게끔 하는 코드를 작성한 후에
마지막으로 리펙토링 과정을 진행했습니다.

Dev-RubinJo committed:
test(step1): TDD를 위한 테스트코드 작성

구현 코드 없이 테스트 코드부터 우선 작성했습니다.
기획에 따라 생각한 객체마다 따로 테스트를 작성했습니다.

Dev-RubinJo committed:
add(step1): 테스트 실패에 따른 테스트 통과 코드 생성

테스트 통과를 위한 코드를 만들고, 범용적인 로직에 맞게 리펙토링했습니다.

YebinKim commented on Lotto/Lotto/StringAddCalculator/Calculator.swift:
```diff
@@ -0,0 +1,24 @@
+        
+        return previous
+    }
+    
+    
```
개행이 2개 추가되어 있네요 👀
의미 없는 개행이라면 삭제해주는 것이 어떨까요?

Dev-RubinJo replied:
어엇 그렇네요. 삭제하겠습니다!

YebinKim commented on Lotto/Lotto/StringAddCalculator/Validator.swift:
```diff
@@ -0,0 +1,34 @@
+        throw ValidatorError.negativeNumber
+    }
+    
+    enum ValidatorError: Error, LocalizedError {
```
`ValidatorError` 는 현재 `Validator` 타입 안에 선언되어 있어 사용할 때 Validator 라는 네이밍이 중복될 것으로 보이네요 🤔
[Swift API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/#promote-clear-usage) 의 **Omit needless words** 에서 설명하는 내용과 완전히 일치하는 맥락은 아니지만, 저는 `ValidatorError` 타입의 Validator 가 작성되지 않아도 충분히 의미전달이 가능할 것으로 보이는데 혹시 유빈님은 어떻게 생각하실까요?

Dev-RubinJo replied:
안그래도 Prefix가 중복되기에 좀 많이 어색한 감이 있었습니다. 하지만 Validator를 제거한 Error로만 하면 Error 예약어를 사용하는 것이기에 우선 그대로 ValidatorError로 넣었던 것인데 다시한번 다른 단어를 사용할 수 있을지 고민해보고 수정하도록 하겠습니다.

YebinKim commented on Lotto/Lotto/StringAddCalculator/Validator.swift:
```diff
@@ -0,0 +1,34 @@
+    enum ValidatorError: Error, LocalizedError {
+        case negativeNumber
+        case invalidNumber
+        var errorDescription: String? {
```
`errorDescription` 에 옵셔널 타입을 활용한 이유는 무엇일까요?

Dev-RubinJo replied:
LocalizedError의 errorDescription이 String? 타입으로 되어있어 그대로 사용했습니다.

YebinKim commented on Lotto/StringAddCalculatorTests/StringAddCalculatorTests.swift:
```diff
@@ -0,0 +1,16 @@
+class StringAddCalculatorTests: XCTestCase {
+    
+    func test_숫자_배열이_주어졌을_때_덧셈이_잘_동작한다 () throws {
+        XCTAssert(Calculator.add([1, 2, 3]) == 6,
```
`XCTAssert` 과 `XCTAssertEqual` 을 혼합해서 사용하신 것으로 보이네요 👀
혹시 어떤 기준으로 구분해서 사용하신 걸까요?

Dev-RubinJo replied:
똑같은 기능을 수행하지만 아무래도 작성하는 방법이 달라서 둘다 연습해보면서 둘다 충분히 손에 익히려고 합니다!

YebinKim commented on Lotto/StringAddCalculatorTests/SeparatorTests.swift:
```diff
@@ -0,0 +1,35 @@
+class SeparatorTests: XCTestCase {
+    
+    func test_빈_문자가_입력이_되면_0을_반환한다() throws {
+        XCTAssertEqual(Separator.separate(nil),
```
현재 작성된 테스트 케이스는 모두 `XCTAssert~` 메서드 안에 직접 값 또는 연산을 넣어주는 형태로 하드코딩 되어있네요!
어떤 값이 주어지는지, 어떤 연산을 수행할 것인지, 기대하는 결과가 무엇인지를 각각 선언하고, 선언한 값을 테스트 메서드에 전달해보면 더욱 읽기 쉬운 테스트 코드가 될 거 같은데 혹시 어떻게 생각하실까요?

Dev-RubinJo replied:
테스트 코드 작성을 시작한지 아직 얼마 안돼서 그정도까지의 생각을 하지 못했습니다. 말해주신 것을 곰곰이 생각을 해보니 기존에 일반 코드들에서도 하드코딩을 지양했는데 테스트코드에서는 그러지 말라는 법이 없군요. 이번부터 한번 최대한 적용해보도록 하겠습니다!

YebinKim commented on Lotto/StringAddCalculatorTests/SeparatorTests.swift:
```diff
@@ -0,0 +1,35 @@
+                       "0이 반환되지 않았습니다.")
+    }
+    
+    func test_컴마_혹은_콜론이_구분자로_입력됐을_때_잘_분리가_된다1() throws {
```
사용하지 않는 throws 키워드가 포함되어 있는 거 같아요! 🤔

Dev-RubinJo replied:
테스트 메서드에서 throws를 사용하지 않을때는 빼도 되는것이었군요! 기본 테스트 메서드에서 다 붙어있어서 넣어줘야하는 것으로 알았습니다. 앞으로도 불필요한 곳이면 빼도록 하겠습니다~

YebinKim commented on Lotto/StringAddCalculatorTests/SeparatorTests.swift:
```diff
@@ -0,0 +1,35 @@
+                       "분리가 정상적으로 이뤄지지 않았습니다.")
+    }
+    
+    func test_컴마_혹은_콜론이_구분자로_입력됐을_때_잘_분리가_된다2() throws {
```
컴마와 콜론이 혼합된 문자열이 잘 분리되는지에 대한 테스트를 수행하면서 각 테스트 메서드의 이름이 1, 2, 3 로 구분된 것으로 보이는데요 👀
테스트 메서드의 이름만 보고 어떤 테스트를 수행하는지 알 수 있도록 네이밍을 구체적으로 변경해보면 어떨까요?

Dev-RubinJo replied:
안그래도 명확하지 않은 느낌이었어서 막연하게 고민중이었는데 다시 더 구체적으로 작성해봐야겠어요.

YebinKim commented on Lotto/Lotto/StringAddCalculator/Converter.swift:
```diff
@@ -0,0 +1,21 @@
+    
+    static func convert(_ numbers: [String]) throws -> [Int] {
+        try numbers.forEach { number in
+            guard try Validator.check(number) else {
+                throw Validator.ValidatorError.negativeNumber
+            }
```
`Validator.check(_:)` 메서드에서 이미 `Validator.ValidatorError.negativeNumber` 를 확인하고 있기 때문에 제가 생각하는 테스트 케이스 중에서는 해당 분기 구문이 실행되지 않을 것으로 보이는데 혹시 유빈님이 생각하신 다른 실행 케이스가 있을까요? 🤔

현재 Code Coverage 에도 포함되지 않는 거 같아 질문드려요~
<img width="866" alt="Converter Code" src="https://user-images.githubusercontent.com/38066039/167287606-e57da79e-ffac-42a6-89a7-1f0d0b1aa1de.png">

Dev-RubinJo replied:
아앗 제가 너무 둘을 나눠서 생각을 해버렸네요. 또한 Converter에서 Validator.ValidatorError를 체크를 하고 있었군요. 이 부분은 다시 검토해서 리펙토링해보겠습니다!

YebinKim commented on Lotto/StringAddCalculatorTests/ValidatorTests.swift:
```diff
@@ -0,0 +1,40 @@
+    }
+    
+    func test_주어진_값에_음수가_있다면_에러를_발생시킨다() throws {
+        XCTAssertThrowsError(try Validator.check("-1") == true,
+                             "숫자가 아닙니다") { error in
+            guard let error = error as? Validator.ValidatorError else {
+                fatalError("ValidatorError가 발생한 것이 아닙니다.")
+            }
+            XCTAssertEqual(error,
+                           Validator.ValidatorError.negativeNumber,
+                           "negativeNumber error가 아닙니다.")
```
하나의 테스트 함수 안에서 실행되는 케이스가 2가지인 것으로 보이네요 👀
_(주어진 값이 음수인지 테스트, 에러의 타입이 `Validator.ValidatorError.negativeNumber` 인지 테스트)_

이 함수 안에서 에러의 타입이 `Validator.ValidatorError` 을 테스트하는 이유가 무엇일까요?

Dev-RubinJo replied:
반환되는 에러가 negativeNumber로 제대로 반환되는지 확인하고자 그렇게 작성했습니다!

YebinKim commented on Lotto/Lotto/StringAddCalculator/Separator.swift:
```diff
@@ -0,0 +1,29 @@
+    }
+}
+
+fileprivate extension String {
```
`fileprivate` 접근제한자를 사용하셨네요!
이 접근제한자를 사용한 이유가 무엇일까요?

Dev-RubinJo replied:
Separator 클래스 내부에서만 해당 extension을 사용하고자 fileprivate를 사용했습니다!

YebinKim commented on Lotto/Lotto/StringAddCalculator/Calculator.swift:
```diff
@@ -0,0 +1,24 @@
+
+import Foundation
+
+final class Calculator {
```
상속 가능성이 없는 클래스에 대해 `final` 키워드가 꼼꼼하게 붙어있네요! 🤩

YebinKim commented on Lotto/Lotto/StringAddCalculator/Calculator.swift:
```diff
@@ -0,0 +1,24 @@
+
+final class Calculator {
+    
+    static func add(_ numbers: [Int]) -> Int {
```
메서드의 이름과 전달인자 레이블, 매개변수 이름이 자연스럽게 읽힌다고 생각합니다!! 😆👍

YebinKim commented on Lotto/Lotto/StringAddCalculator/Validator.swift:
```diff
@@ -0,0 +1,34 @@
+        if num > 0 {
+            return true
+        }
+        throw ValidatorError.negativeNumber
```
요구사항에 따르면 0은 입력 가능한 값임에도 `ValidatorError.negativeNumber` 가 발생하고 있는 것으로 보여요~
해당 케이스 확인 부탁드려요 😊

YebinKim reviewed (changes requested):
유빈님 안녕하세요!
2주차 미션을 함께 진행할 리뷰어 김예빈입니다 😄
서로 소통하며 미션 완주까지 달려보아요!! 🏃🪄

이번 스텝은 요구사항 중 `문자열 계산기에 숫자 이외의 값 또는 음수를 전달하는 경우 Error를 throw 합니다.` 의 상황 외에도 에러가 발생하는 케이스가 있어서 Request Changes 로 리뷰 의견 드렸어요!
작성된 라인별 코멘트도 확인 부탁드립니다 ☺️🙏

또한, 현재 올라온 2개의 커밋은 각각 테스트 케이스 정의 -> 타입 선언 + 내부 로직 작성 의 순서로 진행된 것으로 보이는데요.
중간에 통과만 하게끔 하는 코드도 커밋을 해두면 코드 형상관리에 도움이 되고 아래와 같은 흐름이 더 잘 드러날 수 있을 거 같아요~! 😊

> < TDD 주기 >
> 빨강: 테스트 케이스를 정의 
> 초록: 에러가 발생하지 않도록 필요한 타입을 **껍데기만 작성** (이 과정에서 요구사항에 필요한 타입, 프로퍼티, 메서드 등 생성) 
> 리팩토링: 비즈니스 로직 작성 및 중복 제거

Dev-RubinJo committed:
fix(step1): 피드백 적용하여 문자열 덧셈 계산기 수정

1. Validator의 check 메서드를 throw만 하는 메서드로 변경하여 Converter에서의 불필요 코드 삭제
2. Validator의 ValidatorError 이름에서 Prefix가 중복되고있어 Errors로 수정
3. Tests들에 input, operate, output을 적용하여 보다 가시적으로 볼 수 있도록 수정

YebinKim commented on Lotto/Lotto/StringAddCalculator/Calculator.swift:
```diff
@@ -0,0 +1,22 @@
+final class Calculator {
+    
+    static func add(_ numbers: [Int]) -> Int {
+        var previous: Int = numbers[0]
```
`numbers[0]` 으로 인덱스가 하드코딩되어 있는 것으로 보이네요
`numbers` 가 빈 배열일 경우 out of index 에러가 발생할 가능성이 있지 않을까요? 🤔

YebinKim commented on Lotto/StringAddCalculatorTests/SeparatorTests.swift:
```diff
@@ -0,0 +1,47 @@
+class SeparatorTests: XCTestCase {
+    
+    func test_빈_문자가_입력이_되면_0을_반환한다() {
+        let input: String? = nil
+        let operate: [String] = Separator.separate(input)
+        let output: [String] = ["0"]
```
타입 어노테이션을 꼼꼼히 작성하는 습관 멋지다고 생각합니다! 🤩

YebinKim reviewed (approved):
유빈님 안녕하세요~
리뷰 코멘트 확인 및 반영 고맙습니다! ☺️

이번 스텝에서의 요구사항이 충분히 작성되었고, 메서드는 한 번에 하나의 일만 하고 있으며 메서드 라인수 제한 규칙과 들여쓰기 규칙도 잘 지켜졌다고 생각해서 이번 스텝은 이대로 머지 진행할게요.
Step1 진행 고생 많으셨습니다! 다음 스텝 진행도 화이팅이에요!! 🔥🚀

> 다음 스텝에서는 커밋을 좀 더 나눠서 작성해보면 어떨까요? 코드 형상관리에 도움이 될 거 같아요 👀
